### PR TITLE
Review Notebooks on Data access, chunking and Dask

### PR DESCRIPTION
**Pull request by guillaumeeb**
Creating a draft pull request after the review of the three notebooks on Data and Dask.

Some high level firsts impressions there, and after that I'll try to comment more precisely using github review (or NBReview) later on.

- data_discovery is nice. It lacks a bit of computations or exercise, but it's short so probably OK.
- chunking_introduction is hard. I try to propose some content modifications, but still the part on kerchunk is complicated. Moreover, I'm not sure if Kerchunk is really needed to explain chunks. It's just a good way to consolidate a dataset and make it easily loadable.
- dask_introduction is fine, but it currently lacks real processing (which is perfectly normal as it is probably not finished). I only saw a processing over Lombardia. And as there is currently problem with dask-gateway setup (see https://github.com/pangeo-data/pangeo-eosc/issues/3), it still hard to follow.

### CODE REVIEW

File: tutorial/pangeo101/chunking_introduction.ipynb [hunk 1]
```diff
@@ -49,13 +49,15 @@
    "source": [
```
Comment (guillaumeeb) [inline tutorial/pangeo101/chunking_introduction.ipynb:49]: <p>I'm under the impression that currently, we don't realy talk about compression in this notebook, do we?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/chunking_introduction.ipynb:49]: <p>I added Dask because clearly it is used by Xarray under the hood. I think we should remove some other unsued like pooch?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/chunking_introduction.ipynb:49]: <p>I tried to make things clearer here, but I'm not sure why we are using open_mfdataset right now?</p><p> </p><p>In any case, I think we need to talk about Dask in this notebook. Xarray + chunkings almost always means Dask.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/chunking_introduction.ipynb:49]: <p>Okay, I'm not sure about the last part. I don't know how Xarray chooses chunks when you let it do it automatically. These are not the same chunks as with kerchunk. I think Xarray tries to have chunks bigger enough.</p><p> </p><p>Anyway, Analysis and file reading will still benefit from this chunking.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/chunking_introduction.ipynb:49]: <p>Okay, so what's puzzling me more in this notebook is this part when we directly introduce complex Kerchunk manipulation to access only one file.</p><p> </p><p>I don't know how to do it, but I don't think we need kerchunk to get NetCDF chunks layout, and I'm quite certain that Xarray knows how to optimize this chunk access (but not completly I have to admit).</p><p> </p><p>My current belief is that kerchunk starts to be useful when we really want to open several files without having to read all the metadata in every of those files (or do it only once).</p><p> </p><p>We should at least explain all that a little more, what do you think?</p><p> </p><p>And again, we should probably talk more about chunking, lazyness, partial or sequential processing before introducing kerchunk. I'm not even sure we should show how to create a consolidated dataset with Kerchunk, maybe we should just provide it and point to Kerchunk doc or a misc notebook?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/chunking_introduction.ipynb:49]: <p>Actually, I'm not sure if we want to show content of the file.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

File: tutorial/pangeo101/dask_introduction.ipynb [hunk 2]
```diff
@@ -5,7 +5,7 @@
    "id": "1bfbae7a-12f1-4787-a520-c3de7529168d",
```
Comment (guillaumeeb) [inline tutorial/pangeo101/dask_introduction.ipynb:5]: <p>Maybe here we should talk a bit more about Dask cluster, the role of Client, Scheduler, Workers.</p><p> </p><p>Also talk about the Dashboard and ake people open it to grasp what is going on.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/dask_introduction.ipynb:5]: <p>It's weird we introduce Xarray chunking here while we didn't say a lot on it on the previous notebook.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/dask_introduction.ipynb:5]: <p>I find it really weird that Dask workers do some useless data loading if not optimizing the graph. Are you sure this is really the case if you compute the operation? Usualy, there's nothing more to do when accessing a single element in a collection, Dask knows it only needs to read one chunk.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/dask_introduction.ipynb:5]: <p>This is what I think we should not need. Here we even install basic Pangeo libraries on Dask workers, this is not good. </p><p> </p><p>My opinion is that we should use the same image on both side, with at least Pangeo libraries.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/dask_introduction.ipynb:5]: <p>Maybe we could to some more interesting computation here to show the graph and some tasks executing on the Dashboard?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/dask_introduction.ipynb:5]: <p>Didn't we said we wanted to go global?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

File: tutorial/pangeo101/data_discovery.ipynb [hunk 3]
```diff
@@ -37,19 +37,19 @@
     "    <ul>\n",
```
Comment (guillaumeeb) [inline tutorial/pangeo101/data_discovery.ipynb:37]: <p>General question: do we really need the Setup section in every notebook? </p><p> </p><p>About the package used, should we only talk about those we import, or to which level should we go?</p><p> </p><p>Also, we don't install any package in this notebook.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/data_discovery.ipynb:37]: <p>I added this to show we could get things quickly, but I feel it still lacks a little example computation on the whole dataset.</p><p>And maybe a small conclusion on the fact that Xarray allows concatenating multiple files to form a single dataset, but we have to manually do some operations?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (guillaumeeb) [inline tutorial/pangeo101/data_discovery.ipynb:37]: <p>Not sure if this heading should be kept?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pangeo-data/foss4g-2022/pull/45/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

**Comment by guillaumeeb** [comment]
Okay, I tried to point some interrogations I had about those notebooks.

This is already great work and I know how hard it is to prepare such material, so please don't take my comments badly. I'm only trying to improve this material and makes it as easy as possible for trainees to follow the content.

**Comment by acocac** [comment]
@guillaumeeb all your comments are great. Thanks for your constructive feedback!

**Comment by acocac** [comment]
@guillaumeeb the data and discovery notebook is still, I hope to add a minimal example to access a STAC catalog of Sentinel-2 and maybe another example of available pangeo-forge STAC dataset. I'll merge coming changes in the notebook from the main branch to this PR to get your feedback. Thanks

File: tutorial/pangeo101/data_discovery.ipynb (hunk 3 above)
Comment (pl-marasco) [reply tutorial/pangeo101/data_discovery.ipynb]: <p>The whole dataset will be used once the VCI index is computed. Unfortunatelly this data are already an aggregation of data and there is not too much to play with left. </p><p>Don't forget that there is a general storytelling and some components has been already covered. </p><p>In any case I agree with you, better twice than nothing.</p><p> </p><p>Regarding the concatenation from multiple files isn't too clear to me what you mean with 'we have to manually do some operations?   '</p>

File: tutorial/pangeo101/dask_introduction.ipynb (hunk 2 above)
Comment (pl-marasco) [reply tutorial/pangeo101/dask_introduction.ipynb:5]: <p>You have definetly a point on this but don't forget that is an in persons workshop. Usually some topics are hinted at before the actual explanation.</p><p>What is missing is the new functionality to define the chunks according to a total amount of memory (5mb) that is even better than the auto version. </p>

Comment (pl-marasco) [reply tutorial/pangeo101/dask_introduction.ipynb:5]: <p>One step per time... I would limit the analysis on a small region and, if we have time, we can ask students to scale it up. </p><p>It is also really useful for teaching a good prototyping approach with a second stage of scaling the model to a larger area.&nbsp;</p>

File: tutorial/pangeo101/chunking_introduction.ipynb (hunk 1 above)
Comment (pl-marasco) [reply tutorial/pangeo101/chunking_introduction.ipynb:49]: <p>No I've never touched the topic but definitely is something that would be nice to be added. </p>

**Comment by pl-marasco** [comment]
@guillaumeeb please don't even mention that, without good criticism and different eyes there would not be growth. 

All my material has been prepared from the scratch and on the fly so there is a lot that is missing and is really good to have different visions, opinions and hints.

**Review by annefou — APPROVED** [review]

**Comment by annefou** [comment]
There are some conflicts to fix but apart from that it is ok to me. So once the conflicts are resolved we can merge.

File: tutorial/pangeo101/chunking_introduction.ipynb (hunk 1 above)
Comment (guillaumeeb) [reply tutorial/pangeo101/chunking_introduction.ipynb:49]: <p>We could add some generalities (why compression matters, how it is a gain in processing time despite the overcost of compression, different algorithms currently used in scientific domains, Python blosc, etc.), and maybe a section showing the difference between a NetCDF or zarr file with and witout compression and the resulting size?</p>

File: tutorial/pangeo101/dask_introduction.ipynb (hunk 2 above)
Comment (guillaumeeb) [reply tutorial/pangeo101/dask_introduction.ipynb:5]: <p>I get your point. Anyway, with ma remarks on the chunking notebooks, I think we should show this a bit before (I already added a cell, but it needs more explanation/examples).</p><p> </p>

Comment (guillaumeeb) [reply tutorial/pangeo101/dask_introduction.ipynb:5]: <p>So it's still a work in progress, do you plan to add a section at bigger scale?</p><p> </p><p>You're right it's really important to develop an algorithm and test it at small scale to go bigger later.</p><p> </p><p>I've no idea how much time the processing takes over Lombardia, could we begin at least with the whole Italy?</p>

File: tutorial/pangeo101/data_discovery.ipynb (hunk 3 above)
Comment (guillaumeeb) [reply tutorial/pangeo101/data_discovery.ipynb]: <p>I'm not sure why I put this command there, but anyway, considering the concatenation, I was referring to the preprocess function and the opening of all s3 objects.</p><p> </p><p>This is not really complex, but we've got to do some things in order to present this as a full dataset. This is an interesting introduction to the next notebook and ARCO section.</p>

**Comment by guillaumeeb** [comment]
@annefou I resolved the conflict.

I think it's a god idea to merge before things diverge again.

The thing is I wanted to also discuss some points with all of you (and @tinaok did not made any comment yet). 

So what I propose is that:

- We merge this if everyone agree on the modifications
- I open some issues on the most important points to discuss and solve my comments:
  - Add more explanation on chunking before introducing Kerchunk and multiple file reading. See if we keep all kerchunk part as is. Talk a bit more about Dask arrays.
  - Add some content on compression.
  - Do we keep setup section in all notebooks? It's hard to maintain.
  - Talk more about Dask in dask_introduction, the Dask clusters and dashboard concepts.
  - Remove the part on installing libraries on dask workers (see https://github.com/pangeo-data/pangeo-eosc/issues/3)
  - Scale to the whole world on dask_intro.
- We can continue to discuss little things here if needed.

Thoughts?

**Comment by annefou** [comment]
It is easier to merge and make new issues and PR.
